<ranges>: zip_transform does not accept non const iterable ranges

# Describe the bug
You can't feed ranges which are not const iterable, like `filter_view`, to `zip_transform`.
gcc with libstdc++ accepts this code.

# test case
```
#include <ranges>

int main() {
    using namespace std::views;
    auto evens_and_odds =
        zip_transform([](int even, int odd) { return even + odd; },
                      iota(0, 10) | filter([](int i) { return i % 2 == 0; }),
                      iota(0, 10) | filter([](int i) { return i % 2 != 0; }));
}
```

error is
```
example.cpp
C:/data/msvc/14.38.33133/include\ranges(8872): error C7602: 'std::ranges::range_reference_t': the associated constraints are not satisfied
C:/data/msvc/14.38.33133/include\xutility(2661): note: see declaration of 'std::ranges::range_reference_t'
C:/data/msvc/14.38.33133/include\xutility(2660): note: the concept 'std::ranges::range<const std::ranges::filter_view<std::ranges::iota_view<_Ty1,_Ty2>,main::<lambda_2>>>' evaluated to false
        with
        [
            _Ty1=int,
            _Ty2=int
        ]
C:/data/msvc/14.38.33133/include\xutility(2468): note: call to object of class type 'std::ranges::_Begin::_Cpo': no matching call operator found
C:/data/msvc/14.38.33133/include\xutility(2371): note: could be 'auto std::ranges::_Begin::_Cpo::operator ()(_Ty &&) noexcept(<expr>) const'
C:/data/msvc/14.38.33133/include\ranges(8872): note: the template instantiation context (the oldest one first) is
C:/data/msvc/14.38.33133/include\ranges(8969): note: see reference to variable template 'const bool std::ranges::views::_Zip_transform_fn::_Enable_cpo<`main'::`2'::<lambda_1>,std::ranges::filter_view<std::ranges::iota_view<int,int>,`main'::`2'::<lambda_2> >,std::ranges::filter_view<std::ranges::iota_view<int,int>,`main'::`2'::<lambda_3> > >' being compiled
C:/data/msvc/14.38.33133/include\ranges(8962): note: see reference to class template instantiation 'std::ranges::zip_transform_view<_Func,std::ranges::filter_view<std::ranges::iota_view<_Ty1,_Ty2>,main::<lambda_2>>,std::ranges::filter_view<std::ranges::iota_view<_Ty1,_Ty2>,main::<lambda_3>>>' being compiled
        with
        [
            _Func=main::<lambda_1>,
            _Ty1=int,
            _Ty2=int
        ]
C:/data/msvc/14.38.33133/include\ranges(8872): error C2065: '<missingId>': undeclared identifier
Compiler returned: 2
```

https://godbolt.org/z/o41csG3jr

# Expected behavior
Code compiles without an error

# STL version
Microsoft (R) C/C++ Optimizing Compiler Version 19.38.33133 for x64

--- a/stl/inc/ranges
+++ b/stl/inc/ranges
@@ -8582,6 +8582,10 @@ namespace ranges {
         _EXPORT_STD inline constexpr _Zip_fn zip{};
     } // namespace views
 
+    template <class _Func, class... _ViewTypes>
+    concept _Can_const_iterate_zip_transform_view =
+        range<const zip_view<_ViewTypes...>> && regular_invocable<const _Func&, range_reference_t<const _ViewTypes>...>;
+
     _EXPORT_STD template <class _Func, class... _ViewTypes>
         requires _Zip_transform_constraints<_Func, _ViewTypes...>
     class zip_transform_view : public view_interface<zip_transform_view<_Func, _ViewTypes...>> {
@@ -8833,9 +8837,6 @@ namespace ranges {
             }
         }
 
-        static constexpr bool _Enable_const_begin_end =
-            (range<const _Inner_view> && regular_invocable<const _Func&, range_reference_t<const _ViewTypes>...>);
-
     public:
         zip_transform_view() = default;
 
@@ -8849,7 +8850,7 @@ namespace ranges {
         }
 
         _NODISCARD constexpr auto begin() const noexcept(noexcept(_Iterator<true>{*this, _Zip.begin()})) // strengthened
-            requires _Enable_const_begin_end
+            requires _Can_const_iterate_zip_transform_view<_Func, _ViewTypes...>
         {
             return _Iterator<true>{*this, _Zip.begin()};
         }
@@ -8863,7 +8864,7 @@ namespace ranges {
         }
 
         _NODISCARD constexpr auto end() const noexcept(_Is_end_noexcept<true>()) // strengthened
-            requires _Enable_const_begin_end
+            requires _Can_const_iterate_zip_transform_view<_Func, _ViewTypes...>
         {
             if constexpr (common_range<const _Inner_view>) {
                 return _Iterator<true>{*this, _Zip.end()};

--- a/tests/std/tests/P2321R2_views_zip_transform/test.cpp
+++ b/tests/std/tests/P2321R2_views_zip_transform/test.cpp
@@ -806,6 +806,20 @@ constexpr bool validate_empty_ranges() {
     return true;
 }
 
+// Also test GH-4414: "<ranges>: zip_transform does not accept non const iterable ranges"
+constexpr bool test_gh_4414() {
+    auto evens_and_odds = views::zip_transform([](int even, int odd) { return even + odd; },
+        views::iota(0, 10) | views::filter([](int i) { return i % 2 == 0; }),
+        views::iota(0, 10) | views::filter([](int i) { return i % 2 != 0; }));
+
+    using ZippedTransformed = decltype(evens_and_odds);
+    STATIC_ASSERT(ranges::range<ZippedTransformed>);
+    STATIC_ASSERT(!ranges::range<const ZippedTransformed>);
+
+    constexpr int expected_results[]{1, 5, 9, 13, 17};
+    return ranges::equal(expected_results, evens_and_odds);
+}
+
 int main() {
     // Empty RangeTypes... parameter pack
     {
@@ -857,6 +871,10 @@ int main() {
             test_element_array_one, test_element_array_two, test_element_array_three));
         test_one(three_element_transform_closure, three_range_transform_results_array, test_element_array_one,
             test_element_array_two, test_element_array_three);
+    }
+    {
+        STATIC_ASSERT(test_gh_4414());
+        assert(test_gh_4414());
     }
 
     return 0;

